src/Raven.Server/Documents/Handlers/Admin/AdminIndexHandler.cs

```diff
@@ -49,6 +59,72 @@ public async Task PutJavaScript()
 
             await PutInternal(validatedAsAdmin: false);
         }
+        
+        [RavenAction("/databases/*/indexes/test", "POST", AuthorizationStatus.ValidUser, EndpointType.Write, DisableOnCpuCreditsExhaustion = true)]
+        public async Task TestIndex()
+        {
+            using (ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+            {
+                var input = await context.ReadForMemoryAsync(RequestBodyStream(), "Input");
+
+                var testIndexParameters = JsonDeserializationServer.TestIndexParameters(input);
+                var testIndexDefinition = testIndexParameters.IndexDefinition;
+                var query = testIndexParameters.Query;
+                
+                if (testIndexParameters.IndexDefinition is null)
+                    throw new ArgumentException("Index must have an 'IndexDefinition' field");
+                
+                if (testIndexDefinition.Type.IsJavaScript() == false)
+                {
+                    // C# index without admin authorization
+                    if (HttpContext.Features.Get<IHttpAuthenticationFeature>() is RavenServer.AuthenticateConnection feature && feature.CanAccess(Database.Name, requireAdmin: true, requireWrite: true) == false)
+                        throw new UnauthorizedAccessException($"Index {testIndexDefinition.Name} is a C# index but was sent without admin authorization, this is not allowed.");
+                }
+
+                if (testIndexDefinition.Name is null)
+                    testIndexDefinition.Name = Guid.NewGuid().ToString("N");
+
+                if (query is null)
+                {
+                    var queryString = $"from index '{testIndexDefinition.Name}'";
+                    var queryObject = new { Query = queryString };
+                    
+                    query = DocumentConventions.DefaultForServer.Serialization.DefaultConverter.ToBlittable(queryObject, context);
+                }
+
+                using (var index = Database.IndexStore.CreateTestIndexFromDefinition(testIndexDefinition, context.DocumentDatabase, context))
+                {
+                    index.Start();
+
+                    using (var tracker = new RequestTimeTracker(HttpContext, Logger, Database, "Query"))
+                    using (var token = CreateTimeLimitedQueryToken())
+                    using (var queryContext = QueryOperationContext.Allocate(Database))
+                    {
+                        var indexQueryServerSide = IndexQueryServerSide.Create(HttpContext, query, Database.QueryMetadataCache, tracker, database: Database);
+
+                        indexQueryServerSide.WaitForNonStaleResults = true;
```

Comment:
What will be the value of `WaitForNonStaleResultsTimeout`? Don't we need to specify it?

Reply:
It is set by default to 15 seconds

Reply:
Maybe some configuration option for this? E.g. our ML test index requires downloading hundreds of MB of additional data.

Reply:
I think it's better to have an optional request parameter for this, I added it.